src/webgpu/util/texture/image_copy.ts

```diff
@@ -41,24 +41,33 @@ export function dataBytesForCopy(
     if (offset % info.bytesPerBlock !== 0) valid = false;
     if (layout.bytesPerRow && layout.bytesPerRow % 256 !== 0) valid = false;
   }
-  if (layout.bytesPerRow !== undefined && bytesInLastRow > layout.bytesPerRow) valid = false;
-  if (layout.rowsPerImage !== undefined && heightInBlocks > layout.rowsPerImage) valid = false;
 
   let requiredBytesInCopy = 0;
   {
     let { bytesPerRow, rowsPerImage } = layout;
 
-    // If heightInBlocks > 1, layout.bytesPerRow must be specified.
-    if (heightInBlocks > 1 && bytesPerRow === undefined) valid = false;
-    // If copyExtent.depth > 1, layout.bytesPerRow and layout.rowsPerImage must be specified.
-    if (copyExtent.depth > 1 && rowsPerImage === undefined) valid = false;
-    // If specified, layout.bytesPerRow must be greater than or equal to bytesInLastRow.
-    if (bytesPerRow !== undefined && bytesPerRow < bytesInLastRow) valid = false;
-    // If specified, layout.rowsPerImage must be greater than or equal to heightInBlocks.
-    if (rowsPerImage !== undefined && rowsPerImage < heightInBlocks) valid = false;
-
-    bytesPerRow ??= align(info.bytesPerBlock * widthInBlocks, 256);
-    rowsPerImage ??= heightInBlocks;
+    // (a) If heightInBlocks > 1, layout.bytesPerRow must be specified.
+    // (b) If copyExtent.depth > 1, layout.bytesPerRow and layout.rowsPerImage must be specified.
+    // (c) If specified, layout.bytesPerRow must be greater than or equal to bytesInLastRow.
+    // (d) If specified, layout.rowsPerImage must be greater than or equal to heightInBlocks.
+    //
+    // But for the sake of various tests that don't actually care about the exact value, guess.
+    if (bytesPerRow !== undefined && bytesPerRow < bytesInLastRow) {
```

Comment:
This is structured a bit weird, shouldn't we do all the checks for bytesPerRow (potentially as separate boolean) and then if invalid compute the best guess? Here the state between the ifs makes things a bit complicated.

Reply:
There's still a bit of inbetween state, but I think this is clearer. PTAL.